### ISSUE TITLE
No data in Registry daily around 23:15GMT

### ISSUE THREAD

**Issue by hayfield** [ViderumGlobal/ckanext-iati#112]
On a daily basis there is a period of 15-20 minutes around ~23:15GMT where no data is available in the Registry.

See https://travis-ci.org/IATI/IATI-Website-Tests/builds/192206061 for a failing test from last night. The 23:15GMT test fails on a daily basis. On a small number of occasions, the 23:30GMT test also fails.

Manual inspection of the site at this time shows that the problem is wider than the homepage simply displaying the wrong number (as the test detects):

* The [homepage](https://iatiregistry.org/) displays a `0` in the box stating the number of published Organisations
* The [Publisher page](https://iatiregistry.org/publisher) lists no publishers
* A [search](https://iatiregistry.org/dataset) for all datasets shows no datasets

The CKAN API has not been tested to see whether that returns any data at this time.

**Comment by pmartignoli** [comment]
Hi @hayfield,
We have made some changes to ensure data is available in the registry, also while the nightly script is running. Please get back to us if there are any more issues with this.
Thank you,
Paul

**Comment by hayfield** [comment]
It ran overnight last night without the test failing at all, so this looks to now be sorted - thanks.

**Comment by hayfield** [comment]
There has been a regression, which means this is occurring again (this time around 00:00 GMT).

From last night: [failing test 1](https://travis-ci.org/IATI/IATI-Website-Tests/builds/295545021), [failing test 2](https://travis-ci.org/IATI/IATI-Website-Tests/builds/295548426), [passing test](https://travis-ci.org/IATI/IATI-Website-Tests/builds/295551440).

**Comment by dalepotter** [comment]
@minnaholi I've marked this one as 'high priority', but this is actually quite urgent to resolve once again. A lot of the automated IATI download bots are set to their work around midnight, so they will receive no data at all.  We suspect this may be causing some issues for downstream systems.  For example, the [d-portal](http://d-portal.org/) site is only showing two publishers at present:

![image](https://user-images.githubusercontent.com/8247168/32271004-f8925d66-beef-11e7-9a2a-d7f3ac9af9da.png)

> Hi @hayfield,
We have made some changes to ensure data is available in the registry while the nightly script is running. Please get back to us if there are any more issues with this.
Thank you,
Paul

Once this is cleared up, it would be good to know how this issue has cropped up again since (along with #110) this is an issue raised and fixed some time ago but has now reoccured. This is a good example of [@andylolz's point](https://github.com/ViderumGlobal/ckanext-iati/issues/141#issuecomment-334685013) re documenting bugs to provide context if issues repeat. (Although you might have some details in your internal system??)

**Comment by goranmaxim** [comment]
During maintenance, the reindexing operation was set to run a few hours before it was supposed to, which resulted in this "gap" time where data was unavailable. After purging, the reindexing job runs for ~20 minutes, after which all data should be up to date. We've now set & documented the cron job running times.

**Comment by dalepotter** [comment]
Thanks for the update @goranmaxim we'll keep this issue open for a few days to check that there are no further failing tests.

**Comment by andylolz** [comment]
> During maintenance, the reindexing operation was set to run a few hours before it was supposed to, which resulted in this "gap" time where data was unavailable. After purging, the reindexing job runs for ~20 minutes, after which all data should be up to date. We've now set & documented the cron job running times.

So it sounds like there was a one-off scheduled maintenance task, but before the maintenance was complete, the data was reindexed. Is that right?

**Comment by dalepotter** [comment]
@minnaholi @goranmaxim - this issue appears to be back again in a big way.

Our routine smoke tests have been failing on the Registry since 00:38 UK time
- [Example 1](https://travis-ci.org/IATI/IATI-Website-Tests/builds/296556927) (00:38 - Appears to be ane example of the flakiness described in this issue)
- [Example 2](https://travis-ci.org/IATI/IATI-Website-Tests/builds/296569474) (01:17 - Site was completely down)
- [Example 3](https://travis-ci.org/IATI/IATI-Website-Tests/builds/296670003?utm_source=email&utm_medium=notification) (08:32 - Incorrect number of publishers)

As with example 3, currently the Registry is showing incorrect numbers of publishers:

On https://iatiregistry.org/
![image](https://user-images.githubusercontent.com/8247168/32366624-ca6d6894-c076-11e7-9033-e07a08d0241f.png)

On https://iatiregistry.org/dataset
![image](https://user-images.githubusercontent.com/8247168/32366644-dc651218-c076-11e7-9e74-fcb970b22dee.png)

**Comment by dalepotter** [comment]
The Registry is now showing 183 publishers. 

@minnaholi @goranmaxim - is a rebuild in progress?

![image](https://user-images.githubusercontent.com/8247168/32368119-9a2668ba-c07c-11e7-8cd1-cb0edfae6303.png)

**Comment by JGulic** [comment]
@dalepotter yes we are rebuilding the index, everything should be ok in 15 minutes or so

**Comment by dalepotter** [comment]
Thanks for this update @JGulic

**Comment by hayfield** [comment]
Over the weekend (Saturday, Sunday, Monday), the Registry has been showing 0 publishers (or Datasets) from ~02:00-02:45GMT. This is an even longer period of no data than has been present in previous overnight issues.

**Comment by goranmaxim** [comment]
Since both purging & reindexing take an arbitrary amount of time each time (especially purging, reindexing less so), we can't set them at an exact time and let them run in perfect sync, and be sure they ran without issues each time. They currently run at the [following hours](https://github.com/ViderumGlobal/ckanext-iati#nightly-cronjobs-runtimes).
A little bit of context:
The way normal deletion works, is whatever item is being deleted, it just gets marked as "deleted" in the database, but the entry remains. Purging completely removes the entries from the database. 
The search engine, however, doesn't know about this deletion until a reindexing operation is made. Which is when there is "missing data". 
A suggestion off the top of my head would be to move the purging/reindexing tasks to more suitable hours or make purging a weekly instead of a daily task.

**Comment by hayfield** [comment]
There have been large periods of time where this has not been an issue at all. How was this not a problem at all for many months, and yet is now?

**Comment by goranmaxim** [comment]
Configuration mishap on our end. We've corrected it again so it doesn't clear the index before a rebuild. No data should be missing now, regardless of nightly reindexing.
Thanks @hayfield for pointing us in the right direction.

**Comment by dalepotter** [comment]
Thanks for fixing this one @goranmaxim - once again, I'll mark this issue as 'On hold' open for a few days to check that there are no further failing tests.

Issue #110 was also a reoccurring bug caused by a configuration error - what changes in process are you taking to document such configuration for the future? Am also tagging @minnaholi and @mbocevski as they may be able to offer an insight into your company/development processes too to reassure us.
